hippo_test.go

```diff
@@ -1,90 +1,367 @@
 	}
-	deviceNameBase := strings.Join(devicename, "") + "_%d"
-	r = &TagBatchPart{
+
+	lock := sync.Mutex{}
+	receivedRequests := make([]string, 0)
+
+	// setup test server
+	mux := http.NewServeMux()
+	ts := httptest.NewServer(mux)
+	defer ts.Close()
+	mux.HandleFunc("/kentik/server/url", func(w http.ResponseWriter, r *http.Request) {
+		lock.Lock()
+		defer lock.Unlock()
+
+		t.Helper()
+		serviceCalled = true
+
+		// keep track of the requests received
+		jsonPayload, err := ioutil.ReadAll(r.Body)
+		receivedRequests = append(receivedRequests, string(jsonPayload))
+		a.NoError(err)
+
+		// write the canned response - same for both batches
+		responseBytes, err := json.Marshal(cannedResponse)
+		a.NoError(err)
+
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(200)
+		w.Write(responseBytes)
+	})
+
+	sut := NewHippo("agent", "email", "token")
+	sut.SetSenderInfo("my-service", "service-instance-1", "my-host-name")
+
+	// force the client to use small batches - request as one part is 546 bytes - make the batch size 300
+	sut.OutgoingRequestSize = 300
+
+	// build the request
+	batch := TagBatchPart{
 		ReplaceAll: true,
 		IsComplete: true,
-		Upserts:    make([]TagUpsert, MAX_HIPPO_SIZE/100),
-	}
-
-	for i := 0; i < MAX_HIPPO_SIZE/100; i++ {
-		r.Upserts[i] = TagUpsert{
-			Value: fmt.Sprintf("%d", i),
-			Criteria: []TagCriteria{
-				{
-					Direction:         "dst",
-					DeviceNameRegexes: []string{fmt.Sprintf(deviceNameBase, i)},
+		Upserts: []TagUpsert{
+			{
+				Value: "test1",
+				Criteria: []TagCriteria{
+					{
+						Direction:   "asc",
+						IPAddresses: []string{"1.2.3.4"},
+					},
 				},
 			},
-		}
+			{
+				Value: "test2",
+				Criteria: []TagCriteria{
+					{
+						Direction:   "asc",
+						IPAddresses: []string{"2.2.3.4"},
+					},
+				},
+			},
+			{
+				Value: "test3",
+				Criteria: []TagCriteria{
+					{
+						Direction:   "asc",
+						IPAddresses: []string{"3.2.3.4"},
+					},
+				},
+			},
+			{
+				Value: "test4",
+				Criteria: []TagCriteria{
+					{
+						Direction:   "asc",
+						IPAddresses: []string{"4.2.3.4"},
+					},
+				},
+			},
+			{
+				Value: "test5",
+				Criteria: []TagCriteria{
+					{
+						Direction:   "asc",
+						IPAddresses: []string{"5.2.3.4"},
+					},
+				},
+			},
+		},
+		TTLMinutes: 0,
 	}
 
-	bAllNew, numUp, err := c.EncodeTagBatchPart(r)
-	assert.NoError(t, err)
+	url := fmt.Sprintf("%s/kentik/server/url", ts.URL)
+	response, err := sut.SendBatch(context.Background(), url, &batch)
+	a.NoError(err)
+	a.NotNil(response)
```

Comment:
I don't think I've used require before. Just updated the tests to use it. Good call, thanks.

As far as creating the assert per test - that was just to avoid having to pass `t` in everywhere, that's all. This didn't come from a style guide or anything :)

Reply:
I love you! It's always irritated me a little that these kinds of failures would blow things up (obviously not enough to do anything about it though...)